_posts/2018-08-26-eager-dcgan/eager-dcgan.Rmd

```diff
@@ -0,0 +1,460 @@
+      self$fc1 <- layer_dense(units = 1)
+      
+      function(inputs,
+               mask = NULL,
+               training = TRUE) {
+        inputs %>% self$conv1() %>%
+          self$leaky_relu1() %>%
+          self$dropout(training = training) %>%
+          self$conv2() %>%
+          self$leaky_relu2() %>%
+          self$flatten() %>%
+          self$fc1()
+      }
+    })
+  }
+```
+
+
+## Setting the scene
+
+Before we can start training, we need to create the usual components of a deep learning setup: the model (or models, in this case), the loss function(s), and the optimizer(s).
+
+Model creation is just a function call, with a little extra on top:
+
+```{r}
+generator <- generator()
+discriminator <- discriminator()
+
+# https://www.tensorflow.org/api_docs/python/tf/contrib/eager/defun
+generator$call = tf$contrib$eager$defun(generator$call)
+discriminator$call = tf$contrib$eager$defun(discriminator$call)
+```
+
+[_defun_](https://www.tensorflow.org/api_docs/python/tf/contrib/eager/defun) compiles a Python function (once per different combination of argument shapes and non-tensor objects values)) into a TensorFlow graph, and is used to speed up computations. This comes with side effects and possibly unexpected behavior - please consult the documentation for the details. Here, we were mainly curious in how much of a speedup we might notice when using this from R - in our example, it resulted in a speedup of 130%.
+
+On to the losses. Discriminator loss consists of two parts: Does it correctly identify real images as real, and does it correctly spot fake images as fake.
+Here `real_output` and `generated_output` contain the logits returned from the discriminator - that is, its judgment of whether the respective images are fake or real.
+
+```{r}
+discriminator_loss <- function(real_output, generated_output) {
+  real_loss <- tf$losses$sigmoid_cross_entropy(
+    multi_class_labels = k_ones_like(real_output),
+    logits = real_output)
+  generated_loss <- tf$losses$sigmoid_cross_entropy(
+    multi_class_labels = k_zeros_like(generated_output),
+    logits = generated_output)
+  real_loss + generated_loss
+}
+```
+
+
+Generator loss depends on how the discriminator judged its creations: It would hope for them all to be seen as real. 
+
+```{r}
+generator_loss <- function(generated_output) {
+  tf$losses$sigmoid_cross_entropy(
+    tf$ones_like(generated_output),
+    generated_output)
+}
+```
+
+Now we still need to define optimizers, one for each model.
+
+```{r}
+discriminator_optimizer <- tf$train$AdamOptimizer(1e-4)
+generator_optimizer <- tf$train$AdamOptimizer(1e-4)
+```
+
+
+## Training loop
+
+There are two models, two loss functions and two optimizers, but there is just one training loop, as both models depend on each other. 
+The training loop will be over MNIST images streamed in batches, but we still need input to the generator - a random vector of size 100, in this case.
+
+```{r}
+noise_dim <- 100L
```

Comment:
Let's make sure we need the `L` suffix here

Reply:
you're right... probably I had tf$random_normal here before...